NixOS module configuration improvements

quentinmit opened the pull request:
This branch has three different configuration improvements.

The first is that it disables user creation when no users are configured. This is important because otherwise the `refresh_users.py` script will try to delete all the users in the database. (I'm actually not sure how this could work for anyone, because once you actually create parts in the database, foreign keys will prevent the delete + recreate that `refresh_users.py` will try to do.)

With this change, it's possible to configure SSO login:

```nix
services.inventree.config = {
  social_backends = [
    "allauth.socialaccount.providers.openid_connect"
  ];
  global_settings.LOGIN_ENABLE_SSO = true;
  global_settings.LOGIN_ENABLE_SSO_REG = true;
  global_settings.LOGIN_ENABLE_PWD_FORGOT = false;
};
```

The second improvement is to restore the recently-deleted `serverBind` option; it was replaced with `bindIp` and `bindPort` settings, but the latter cannot be used to specify a Unix socket such as:

```nix
services.inventree = {
  serverBind = "unix:/run/inventree/inventree.sock";
}
```

Finally, I replaced the `siteUrl` and `allowedHosts` settings with `config.site_url` and `config.allowed_hosts` respectively. Having the settings as environment variables was causing the default values to override the explicit settings I had put in `config`; now, they are only set in one place. (I used `lib.mkRenamedOptionModule` to not break any existing configurations and allow the old option name to continue working.)

quentinmit committed:
Fix support for configuring a Unix socket listener

quentinmit committed:
Set site_url and allowed_hosts via the config file

oddlama commented:
Any status updates on this one? I would love to use this to enable SSO on my instance.

Also I have a suggestion for a small improvement: Instead of

```nix
configFile = configFormat.generate "config.yaml" cfg.config;
```

you may want to have a look at `utils.genJqSecretsReplacementSnippet` from nixpkgs which allows specifying secrets in the yaml file at any point using `somevalue._secret = "/path/to/file"`. That would make it pretty easy to specify smtp or database passwords in the freeform settings. [Usage example](https://github.com/NixOS/nixpkgs/blob/d0cc7d8f47e12841faef3203d79893c6396397a2/nixos/modules/services/web-apps/ente.nix#L209)

Gigahawk commented:
I mean I'm not going to merge it while there are merge conflicts, and quite frankly I'm not really sure how I feel about being able to declare other DB settings if values that aren't specified don't get return to default by the script.

quentinmit commented:
> I mean I'm not going to merge it while there are merge conflicts, and quite frankly I'm not really sure how I feel about being able to declare other DB settings if values that aren't specified don't get return to default by the script.

@Gigahawk You force-pushed this PR four times, so my assumption was that you were taking it over to merge it. I'm happy to rebase it if you'd like - just let me know.

It looks like InvenTree [now](https://github.com/inventree/InvenTree/commit/4b71130ebed3d634dd5da05b0037f96d036a9a66) has a `INVENTREE_GLOBAL_SETTINGS` environment variable that it pays attention to. I can refactor this PR to set that environment variable instead.

quentinmit committed:
Skip user creation if no users are configured

quentinmit commented:
> > I mean I'm not going to merge it while there are merge conflicts, and quite frankly I'm not really sure how I feel about being able to declare other DB settings if values that aren't specified don't get return to default by the script.
> 
> @Gigahawk You force-pushed this PR four times, so my assumption was that you were taking it over to merge it. I'm happy to rebase it if you'd like - just let me know.
> 
> It looks like InvenTree [now](https://github.com/inventree/InvenTree/commit/4b71130ebed3d634dd5da05b0037f96d036a9a66) has a `INVENTREE_GLOBAL_SETTINGS` environment variable that it pays attention to. I can refactor this PR to set that environment variable instead.

@Gigahawk I've rebased this PR and removed the settings bit. I've updated the first comment with a demo of how you can configure SSO with `services.inventree.config.global_settings.LOGIN_ENABLE_SSO`. This PR is still necessary to prevent SSO users from being deleted at next startup.

PTAL

quentinmit commented:
> Also I have a suggestion for a small improvement: Instead of
> 
> ```nix
> configFile = configFormat.generate "config.yaml" cfg.config;
> ```
> 
> you may want to have a look at `utils.genJqSecretsReplacementSnippet` from nixpkgs which allows specifying secrets in the yaml file at any point using `somevalue._secret = "/path/to/file"`. That would make it pretty easy to specify smtp or database passwords in the freeform settings. [Usage example](https://github.com/NixOS/nixpkgs/blob/d0cc7d8f47e12841faef3203d79893c6396397a2/nixos/modules/services/web-apps/ente.nix#L209)

@oddlama Sorry I missed this suggestion before. I've been using environment variables to pass in my secrets:

```nix
  sops.templates."inventree.env".content = ''
    INVENTREE_SOCIAL_PROVIDERS=${lib.strings.toJSON (lib.strings.toJSON {
      openid_connect.APPS = [{
        provider_id = "foo";
        name = "Foo";
        client_id = config.sops.placeholder."authentik/apps/inventree/client_id";
        secret = config.sops.placeholder."authentik/apps/inventree/client_secret";
        settings.server_url = "https://auth.foo.com/application/o/inventree/.well-known/openid-configuration";
        settings.token_auth_method = "client_secret_basic";
      }];
    })}
  '';
  systemd.services.inventree-server.serviceConfig.EnvironmentFile = config.sops.templates."inventree.env".path;
```

I'm not quite sure how to use `genJqSecretsReplacement` or `genJqSecretsReplacementSnippet`. The former seems to be new as of 25.11 (and I see your username in the blame, ha!).

Unless I'm missing something, `genJqSecretsReplacement*` don't seem to actually set the mode of the file it creates. Wouldn't that result in a world-readable file? Also, the config file is currently being generated as part of `ExecStartPre`, which is being run as root. I don't know why that would be, either - it only seems to touch files inside `/var/lib/inventree` that should be owned by the `inventree` user.

And what is the correct option type for an option that takes something that will end up going through `genJqSecretsReplacement`? I guess `(pkgs.formats.json {}).type` as I'm doing here, even though we won't use the corresponding `generate` method?

I think I would want to peel this off into its own PR so resolving these questions doesn't block the other changes here. I agree it would be a nice improvement, though.

Gigahawk commented:
LGTM, thanks for rebasing.
